storm-server/src/main/java/org/apache/storm/metric/timed/TimerDecorated.java

```diff
@@ -0,0 +1,47 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+
+package org.apache.storm.metric.timed;
+
+import com.codahale.metrics.Timer;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+public interface TimerDecorated extends AutoCloseable {
+
+    boolean hasStopped();
+
+    default boolean hasStopped(final AtomicReference<Timer.Context> timing) {
+        return timing.get() == null;
+    }
+
+    long stopTiming();
```

Comment:
I am a little concerned that we have both close and stopTiming, but if anyone calls stopTiming and then close the assertion will be thrown.  It is like you have a method that you want everyone to implement, but never use.

Could you add some javadocs to this method in particular, but also to the others as well so that developers understand how to use these methods, as it is not the most obvious.  It would also be nice if the javadocs could point to one of the implementations as an example of how to extend this as a cross cutting concern.

Reply:
I'm actually modeling so after the implementation of dropwizard's Timer class. Basically the object can behave like an Autocloseable and put in try-with-resource block or specialized collections (such as TimeCacheMap). In a way, close method is not for users to call and user should either invoke stop timing or rely on `Autocloseable`'s behavior. But yeah now I think for robustness maybe `stopTiming` should be idempotent instead of only a one time call.

Reply:
Also I'm not sure if the current design is best for extensibility. I can think of a scenario when you want to have different timer or track time in different phase (say you want to measure not only how long it takes to launch an assignment but how long it runs on a worker). Should I make it more flexible so it's easier to maintain in the future?